packages/astro/src/config/types.ts

```diff
@@ -0,0 +1,30 @@
+import type fsMod from 'node:fs';
+import type { AstroSettings, InjectedDts } from '../@types/astro.js';
+import { AstroError, AstroErrorData } from '../core/errors/index.js';
+
+// TODO: only keep astro.d.ts in Astro 5
+export const CODEGENDIR_BASE_DTS_FILE = 'types.d.ts';
+const RESERVED_FILE_NAMES = ['astro.d.ts', CODEGENDIR_BASE_DTS_FILE];
+
+export function injectDts({
+	codegenDir,
+	filename,
+	content,
+	bypassValidation = false,
+	fs,
+}: Pick<AstroSettings, 'codegenDir'> &
+	InjectedDts & { bypassValidation?: boolean; fs: typeof fsMod }) {
+	if (
+		!bypassValidation &&
+		(!filename.endsWith('.d.ts') || RESERVED_FILE_NAMES.includes(filename))
+	) {
+		throw new AstroError(AstroErrorData.InvalidInjectTypesFilename);
+	}
+
+	ensureCodegenDirExists({ codegenDir, fs });
+	fs.writeFileSync(new URL(filename, codegenDir), content);
```

Comment:
I/O operations are expensive, and we should minimise them whenever possible. 

This means that `injectDts` is going write to disk every single time we call it. Instead, we could save the contents in some internal object and do writing operations at the very end.

The object could be something like:
```js
const dts = {
	["<INTEGRATION_NAME>"]: ["contentA", "contentB"]
}
```

Reply:
Actually, injectDts is only called in 2 places:
1. At the end of syncInternal, where we loop over each injected dts (both core and integrations)
2. When calling `createContentTypesGenerator`. That one is a bit unfortunate but basically, `injectDts` is called by this function after we write the `types.d.ts` file with the references. To work around this, We first inject an empty `content.d.ts` so that its part of the references and then, we call `injectDts` whenever we want in the content plugin

That's truly the only places we're we doing file writes.

However, calling `ensureCodegenDirExists` each time may not be optimal. The thing, I spent quite some time debugging and probably because I'm misunderstanding in what order parts run, I could only fix the issue by calling this function in there :/. Ideally we'd basically call it in 1 place and call it a day but I couldn't manage to find where.